src/lib/parsers/sol-json.c

```diff
@@ -607,62 +607,83 @@ sol_json_calculate_escaped_string_len(const char *str)
 }
 
 SOL_API char *
-sol_json_escape_string(const char *str, char *buf, size_t len)
+sol_json_escape_string(const char *str, struct sol_buffer *buf)
 {
-    char *out = buf;
-    size_t i;
+    char *out, *r_str;
+    size_t i, escaped_len, new_size;
+    int r;
 
     SOL_NULL_CHECK(str, NULL);
     SOL_NULL_CHECK(buf, NULL);
 
-    for (i = 0; *str && i < len; str++, i++) {
+    escaped_len = sol_json_calculate_escaped_string_len(str);
+
+    r = sol_util_size_add(buf->used, escaped_len, &new_size);
+    SOL_INT_CHECK(r, < 0, NULL);
+
+    r = sol_buffer_ensure(buf, new_size);
+    SOL_INT_CHECK(r, < 0, NULL);
+
+    r_str = out = sol_buffer_at_end(buf);
+
+    for (i = 0; *str && i < escaped_len; str++, i++) {
         if (memchr(sol_json_escapable_chars, *str, sizeof(sol_json_escapable_chars))) {
-            *buf++ = '\\';
+            *out++ = '\\';
             switch (*str) {
-            case '"':  *buf++ = '"'; break;
-            case '\\': *buf++ = '\\'; break;
-            case '/':  *buf++ = '/'; break;
-            case '\b': *buf++ = 'b'; break;
-            case '\f': *buf++ = 'f'; break;
-            case '\n': *buf++ = 'n'; break;
-            case '\r': *buf++ = 'r'; break;
-            case '\t': *buf++ = 't'; break;
+            case '"':  *out++ = '"'; break;
+            case '\\': *out++ = '\\'; break;
+            case '/':  *out++ = '/'; break;
+            case '\b': *out++ = 'b'; break;
+            case '\f': *out++ = 'f'; break;
+            case '\n': *out++ = 'n'; break;
+            case '\r': *out++ = 'r'; break;
+            case '\t': *out++ = 't'; break;
             }
         } else {
-            *buf++ = *str;
+            *out++ = *str;
         }
     }
-    *buf++ = '\0';
-    return out;
+
+    *out++ = '\0';
+    buf->used += escaped_len - 1;
+
+    return r_str;
 }
 
 SOL_API int
-sol_json_double_to_str(const double value, char *buf, size_t len)
+sol_json_double_to_str(const double value, struct sol_buffer *buf)
 {
     int ret;
-    char *decimal_point;
+    char *decimal_point, *end;
+    size_t len;
 
 #ifdef HAVE_LOCALE
     struct lconv *lc = localeconv();
 #endif
 
     SOL_NULL_CHECK(buf, -EINVAL);
+    len = buf->capacity - buf->used;
```

Comment:
as before, this won't work for empty strings. And you should be using `sol_buffer_append_printf()` below :-)

Review what could be simplified with `sol-buffer.h`, eventually even the memmove can go.